core/src/main/java/com/linecorp/armeria/internal/common/AbstractSplitHttpMessage.java

```diff
@@ -0,0 +1,113 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.internal.common;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.concurrent.CompletableFuture;
+
+import org.reactivestreams.Subscriber;
+
+import com.linecorp.armeria.common.HttpData;
+import com.linecorp.armeria.common.HttpHeaders;
+import com.linecorp.armeria.common.HttpMessage;
+import com.linecorp.armeria.common.SplitHttpMessage;
+import com.linecorp.armeria.common.stream.StreamMessage;
+import com.linecorp.armeria.common.stream.SubscriptionOption;
+import com.linecorp.armeria.internal.common.stream.NoopSubscription;
+
+import io.netty.util.concurrent.EventExecutor;
+
+abstract class AbstractSplitHttpMessage implements SplitHttpMessage, StreamMessage<HttpData> {
+
+    private final HttpMessage upstream;
+    private final EventExecutor upstreamExecutor;
+    private final BodySubscriber bodySubscriber;
+
+    protected AbstractSplitHttpMessage(HttpMessage upstream, EventExecutor executor,
+                                       BodySubscriber bodySubscriber) {
+        this.upstream = requireNonNull(upstream, "upstream");
+        upstreamExecutor = requireNonNull(executor, "executor");
+        this.bodySubscriber = bodySubscriber;
+        upstream.subscribe(bodySubscriber, upstreamExecutor, SubscriptionOption.values());
+    }
+
+    @Override
+    public final StreamMessage<HttpData> body() {
+        return this;
+    }
+
+    @Override
+    public final CompletableFuture<HttpHeaders> trailers() {
+        return bodySubscriber.trailersFuture();
+    }
+
+    @Override
+    public final boolean isOpen() {
+        return upstream.isOpen();
+    }
+
+    @Override
+    public final boolean isEmpty() {
+        return !isOpen() && !bodySubscriber.wroteAny();
+    }
+
+    @Override
+    public final long demand() {
+        return upstream.demand();
+    }
+
+    @Override
+    public final CompletableFuture<Void> whenComplete() {
+        return upstream.whenComplete();
+    }
+
+    @Override
+    public final void abort() {
+        upstream.abort();
+    }
+
+    @Override
+    public final void abort(Throwable cause) {
+        upstream.abort(cause);
+    }
+
+    @Override
+    public final EventExecutor defaultSubscriberExecutor() {
+        return upstreamExecutor;
+    }
+
+    @Override
+    public void subscribe(Subscriber<? super HttpData> subscriber, EventExecutor executor,
+                          SubscriptionOption... options) {
+        requireNonNull(subscriber, "subscriber");
+        requireNonNull(executor, "executor");
+        requireNonNull(options, "options");
+
+        if (!BodySubscriber.downstreamUpdater().compareAndSet(bodySubscriber, null, subscriber)) {
```

Comment:
I guess my comment was not clear. 😅 
What I meant was using `subscribedUpdater` to allow only one subscriber:
```java
abstract class AbstractSplitHttpMessage implements SplitHttpMessage, StreamMessage<HttpData> {

    private static final AtomicIntegerFieldUpdater<AbstractSplitHttpMessage> subscribedUpdater =
            AtomicIntegerFieldUpdater.newUpdater(AbstractSplitHttpMessage.class, "subscribed");

    private volatile int subscribed;
    
    ...

    @Override
    public void subscribe(Subscriber<? super HttpData> subscriber, EventExecutor executor,
                          SubscriptionOption... options) {
        requireNonNull(subscriber, "subscriber");
        requireNonNull(executor, "executor");
        requireNonNull(options, "options");

        if (!subscribedUpdater.compareAndSet(this, 0, 1)) {
            subscriber.onSubscribe(NoopSubscription.get());
            subscriber.onError(new IllegalStateException("Only single subscriber is allowed!"));
            return;
        }

        if (executor.inEventLoop()) {
            bodySubscriber.initDownstream(subscriber, executor, options);
        } else {
            executor.execute(() -> bodySubscriber.initDownstream(subscriber, executor, options));
        }
    }
}
```
And we don't need the `downstreamUpdater` because we can just assign the value in the `initDownstream` method. :wink:
```java
class BodySubscriber implements Subscriber<HttpObject>, Subscription {
    protected void initDownstream(Subscriber<? super HttpData> downstream, EventExecutor executor,
                                  SubscriptionOption... options) {
        this.downstream = downstream;
        ...
    }
}
```

Reply:
@minwoox, Oh, I see. I misunderstood.
As per your comment, I removed the downstreamUpdater. 👍 
Thank you! 😄